File: src/ML2Typescript/Examples/TS/test8.ts
import { Option, none, some } from './Lib.js';
import * as ArrayList from './ArrayList.js';

function map<B, A>(f: (_: A) => B, l: A[]): B[] {
    
    function map2(acc: any, l: any): any {
        if (l.length == 0) {
            return acc;
        } else {
            const [hd, ...tl] = l;
            return map2([f(hd), ...acc], tl);
        }
    }
    return map2([], l);
}

File: src/ML2Typescript/Examples/ML/test8.ml

let rec map (f: 'a -> 'b) (l: 'a list): 'b list =
  
  let rec map2 acc l =
    match l with
    | [] -> 
      acc
    | (hd :: tl) -> 
      (map2 ((f hd) :: acc) tl)
    
  in
  (map2 [] l);;
